src/pip/_internal/network/auth.py

```diff
@@ -266,27 +265,57 @@ def _get_new_credentials(
                 logger.debug("Found credentials in index url for %s", netloc)
                 return index_url_user_password
 
-        # Get creds from netrc if we still don't have them
-        if allow_netrc:
-            netrc_auth = get_netrc_auth(original_url)
-            if netrc_auth:
-                logger.debug("Found credentials in netrc for %s", netloc)
-                return netrc_auth
-
-        # If we don't have a password and keyring is available, use it.
-        if allow_keyring:
-            # The index url is more specific than the netloc, so try it first
-            # fmt: off
-            kr_auth = (
-                get_keyring_auth(index_url, username) or
-                get_keyring_auth(netloc, username)
-            )
-            # fmt: on
+        netrc_auth = get_netrc_auth(original_url)
+        if netrc_auth is not None:
+            logger.debug("Found credentials in netrc for %s", netloc)
+            return netrc_auth
+
+        kr_auth = self._find_key_ring_credentials(
+            index_url, index_url_user_password, netloc, username
+        )
+        if kr_auth is not None:
+            return kr_auth
+
+        return username, password
+
+    def _find_key_ring_credentials(
+        self,
+        index_url: Optional[str],
+        index_url_user_password: Optional[str],
+        netloc: str,
+        artifact_username: str,
+    ) -> Optional[AuthInfo]:
+        def get_key_ring_user() -> Optional[str]:
+            if artifact_username is not None:
+                return artifact_username
+            if index_url_user_password:
+                if (
+                    index_url_user_password[0] is not None
+                    and index_url_user_password[1] is None
+                ):
+                    logger.debug("Found key ring username in index_url")
+                    return index_url_user_password[0]
+            if artifact_username is None and self.default_key_ring_user is not None:
+                logger.debug("Using default_key_ring_user")
+                return self.default_key_ring_user
+            return None
+
+        key_ring_user = get_key_ring_user()
+        if key_ring_user is None:
+            return None
+
+        if index_url is not None:
+            kr_auth = get_keyring_auth(index_url, key_ring_user)
             if kr_auth:
-                logger.debug("Found credentials in keyring for %s", netloc)
+                logger.debug("Found credentials in keyring for %s", index_url)
                 return kr_auth
 
-        return username, password
+        kr_auth = get_keyring_auth(netloc, key_ring_user)
```

Comment:
Maybe here we should change the order and first try to take credentials for some particular netloc.